# Trabajo Preparacion, visualizacion de Datos y Machine learning con Python<a class="tocSkip">
## Ciencia de datos en Produccion <a class="tocSkip">

**Estudiante:** Maria Alejandra Caicedo Chaves

**ID:** 1016110796

**Email:** macaicedoc_1@uqvirtual.edu.co

**Estudiante:** Iván Andrés Guapacha Barrero

**ID:** 1096040095

**Email:** iaguapachab@uqvirtual.edu.co
 
**Docente:** [Jose R. Zapata](https://joserzapata.github.io)
- https://joserzapata.github.io
- https://twitter.com/joserzapata
- https://www.linkedin.com/in/jose-ricardo-zapata-gonzalez/       

<h1>Contenido<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Definir-el-Problema-a-Resolver" data-toc-modified-id="Definir-el-Problema-a-Resolver-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Definir el Problema a Resolver</a></span><ul class="toc-item"><li><span><a href="#Describir-los-datos-de-entrada-y-salida" data-toc-modified-id="Describir-los-datos-de-entrada-y-salida-1.1"><span class="toc-item-num">1.1&nbsp;&nbsp;</span>Describir los datos de entrada y salida</a></span></li></ul></li><li><span><a href="#Importar-Librerias" data-toc-modified-id="Importar-Librerias-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Importar Librerias</a></span></li><li><span><a href="#Cargar-Datasets" data-toc-modified-id="Cargar-Datasets-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Cargar Datasets</a></span></li><li><span><a href="#Descripcion--y-Limpieza-de-los-datos" data-toc-modified-id="Descripcion--y-Limpieza-de-los-datos-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Descripcion  y Limpieza de los datos</a></span><ul class="toc-item"><li><span><a href="#Identificacion-de-Variables" data-toc-modified-id="Identificacion-de-Variables-4.1"><span class="toc-item-num">4.1&nbsp;&nbsp;</span>Identificacion de Variables</a></span></li><li><span><a href="#Analisis-General-Univariable-y-Bivariable" data-toc-modified-id="Analisis-General-Univariable-y-Bivariable-4.2"><span class="toc-item-num">4.2&nbsp;&nbsp;</span>Analisis General Univariable y Bivariable</a></span></li><li><span><a href="#Eliminar-columnas-de-datos-Innecesarios" data-toc-modified-id="Eliminar-columnas-de-datos-Innecesarios-4.3"><span class="toc-item-num">4.3&nbsp;&nbsp;</span>Eliminar columnas de datos Innecesarios</a></span></li><li><span><a href="#Remover-Datos-Duplicados-Exactos" data-toc-modified-id="Remover-Datos-Duplicados-Exactos-4.4"><span class="toc-item-num">4.4&nbsp;&nbsp;</span>Remover Datos Duplicados Exactos</a></span></li><li><span><a href="#Procesamiento-de-Datos-Faltantes" data-toc-modified-id="Procesamiento-de-Datos-Faltantes-4.5"><span class="toc-item-num">4.5&nbsp;&nbsp;</span>Procesamiento de Datos Faltantes</a></span><ul class="toc-item"><li><span><a href="#Borrar-Filas" data-toc-modified-id="Borrar-Filas-4.5.1"><span class="toc-item-num">4.5.1&nbsp;&nbsp;</span>Borrar Filas</a></span></li><li><span><a href="#Reemplazar-datos-faltantes-con-la-Media/-Moda/-Mediana-(Mean/-Mode/-Median-Imputation)" data-toc-modified-id="Reemplazar-datos-faltantes-con-la-Media/-Moda/-Mediana-(Mean/-Mode/-Median-Imputation)-4.5.2"><span class="toc-item-num">4.5.2&nbsp;&nbsp;</span>Reemplazar datos faltantes con la Media/ Moda/ Mediana (Mean/ Mode/ Median Imputation)</a></span></li></ul></li><li><span><a href="#Remover-Datos-Duplicados-Exactos" data-toc-modified-id="Remover-Datos-Duplicados-Exactos-4.6"><span class="toc-item-num">4.6&nbsp;&nbsp;</span>Remover Datos Duplicados Exactos</a></span></li><li><span><a href="#Analisis-Univariable" data-toc-modified-id="Analisis-Univariable-4.7"><span class="toc-item-num">4.7&nbsp;&nbsp;</span>Analisis Univariable</a></span><ul class="toc-item"><li><span><a href="#Variables-Numericas" data-toc-modified-id="Variables-Numericas-4.7.1"><span class="toc-item-num">4.7.1&nbsp;&nbsp;</span>Variables Numericas</a></span></li><li><span><a href="#Variables-Categoricas" data-toc-modified-id="Variables-Categoricas-4.7.2"><span class="toc-item-num">4.7.2&nbsp;&nbsp;</span>Variables Categoricas</a></span></li></ul></li><li><span><a href="#Analisis-Bivariable" data-toc-modified-id="Analisis-Bivariable-4.8"><span class="toc-item-num">4.8&nbsp;&nbsp;</span>Analisis Bivariable</a></span><ul class="toc-item"><li><span><a href="#Numericas-vs-Numericas" data-toc-modified-id="Numericas-vs-Numericas-4.8.1"><span class="toc-item-num">4.8.1&nbsp;&nbsp;</span>Numericas vs Numericas</a></span></li><li><span><a href="#Categoricas-vs-Categoricas" data-toc-modified-id="Categoricas-vs-Categoricas-4.8.2"><span class="toc-item-num">4.8.2&nbsp;&nbsp;</span>Categoricas vs Categoricas</a></span></li><li><span><a href="#Categoricas-vs-Numericas" data-toc-modified-id="Categoricas-vs-Numericas-4.8.3"><span class="toc-item-num">4.8.3&nbsp;&nbsp;</span>Categoricas vs Numericas</a></span></li></ul></li><li><span><a href="#Procesamiento-de-Outliers" data-toc-modified-id="Procesamiento-de-Outliers-4.9"><span class="toc-item-num">4.9&nbsp;&nbsp;</span>Procesamiento de Outliers</a></span><ul class="toc-item"><li><span><a href="#Deteccion-de-Outliers-(Univariables-y-Bi-variables)" data-toc-modified-id="Deteccion-de-Outliers-(Univariables-y-Bi-variables)-4.9.1"><span class="toc-item-num">4.9.1&nbsp;&nbsp;</span>Deteccion de Outliers (Univariables y Bi variables)</a></span></li><li><span><a href="#Remover-Outliers" data-toc-modified-id="Remover-Outliers-4.9.2"><span class="toc-item-num">4.9.2&nbsp;&nbsp;</span>Remover Outliers</a></span></li></ul></li><li><span><a href="#Procesamiento-de-Outliers" data-toc-modified-id="Procesamiento-de-Outliers-4.10"><span class="toc-item-num">4.10&nbsp;&nbsp;</span>Procesamiento de Outliers</a></span><ul class="toc-item"><li><span><a href="#Deteccion-de-Outliers-(Univariables-y-Bi-variables)" data-toc-modified-id="Deteccion-de-Outliers-(Univariables-y-Bi-variables)-4.10.1"><span class="toc-item-num">4.10.1&nbsp;&nbsp;</span>Deteccion de Outliers (Univariables y Bi variables)</a></span></li><li><span><a href="#Remover-Outliers" data-toc-modified-id="Remover-Outliers-4.10.2"><span class="toc-item-num">4.10.2&nbsp;&nbsp;</span>Remover Outliers</a></span></li></ul></li><li><span><a href="#Feature-Engineering" data-toc-modified-id="Feature-Engineering-4.11"><span class="toc-item-num">4.11&nbsp;&nbsp;</span>Feature Engineering</a></span><ul class="toc-item"><li><span><a href="#Transformacion-de-Variables" data-toc-modified-id="Transformacion-de-Variables-4.11.1"><span class="toc-item-num">4.11.1&nbsp;&nbsp;</span>Transformacion de Variables</a></span><ul class="toc-item"><li><span><a href="#Normalizacion" data-toc-modified-id="Normalizacion-4.11.1.1"><span class="toc-item-num">4.11.1.1&nbsp;&nbsp;</span>Normalizacion</a></span></li><li><span><a href="#Escalamiento" data-toc-modified-id="Escalamiento-4.11.1.2"><span class="toc-item-num">4.11.1.2&nbsp;&nbsp;</span>Escalamiento</a></span></li><li><span><a href="#Logaritmica" data-toc-modified-id="Logaritmica-4.11.1.3"><span class="toc-item-num">4.11.1.3&nbsp;&nbsp;</span>Logaritmica</a></span></li><li><span><a href="#raíz-cuadrada-/-cúbica" data-toc-modified-id="raíz-cuadrada-/-cúbica-4.11.1.4"><span class="toc-item-num">4.11.1.4&nbsp;&nbsp;</span>raíz cuadrada / cúbica</a></span></li><li><span><a href="#Binning-,-Cambios-de-Numericas-a-Categoricas" data-toc-modified-id="Binning-,-Cambios-de-Numericas-a-Categoricas-4.11.1.5"><span class="toc-item-num">4.11.1.5&nbsp;&nbsp;</span>Binning , Cambios de Numericas a Categoricas</a></span></li></ul></li></ul></li><li><span><a href="#Analisis-Univariable-y-Bivariable-Final" data-toc-modified-id="Analisis-Univariable-y-Bivariable-Final-4.12"><span class="toc-item-num">4.12&nbsp;&nbsp;</span>Analisis Univariable y Bivariable Final</a></span><ul class="toc-item"><li><span><a href="#Creacion-de-Variables" data-toc-modified-id="Creacion-de-Variables-4.12.1"><span class="toc-item-num">4.12.1&nbsp;&nbsp;</span>Creacion de Variables</a></span><ul class="toc-item"><li><span><a href="#Crear-Variables-derivadas-de-Otras" data-toc-modified-id="Crear-Variables-derivadas-de-Otras-4.12.1.1"><span class="toc-item-num">4.12.1.1&nbsp;&nbsp;</span>Crear Variables derivadas de Otras</a></span></li><li><span><a href="#Crear-Variables-de-Categorico-a-Numerico" data-toc-modified-id="Crear-Variables-de-Categorico-a-Numerico-4.12.1.2"><span class="toc-item-num">4.12.1.2&nbsp;&nbsp;</span>Crear Variables de Categorico a Numerico</a></span></li></ul></li></ul></li><li><span><a href="#Reduccion-de-Dimensionalidad-y-Seleccion-de-Variables-(PCA)" data-toc-modified-id="Reduccion-de-Dimensionalidad-y-Seleccion-de-Variables-(PCA)-4.13"><span class="toc-item-num">4.13&nbsp;&nbsp;</span>Reduccion de Dimensionalidad y Seleccion de Variables (PCA)</a></span></li><li><span><a href="#Balance-de-datos" data-toc-modified-id="Balance-de-datos-4.14"><span class="toc-item-num">4.14&nbsp;&nbsp;</span>Balance de datos</a></span></li></ul></li><li><span><a href="#MODELAMIENTO-DE-LOS-DATOS-(MACHINE-LEARNING)" data-toc-modified-id="MODELAMIENTO-DE-LOS-DATOS-(MACHINE-LEARNING)-5"><span class="toc-item-num">5&nbsp;&nbsp;</span>MODELAMIENTO DE LOS DATOS (MACHINE LEARNING)</a></span><ul class="toc-item"><li><span><a href="#Dividir-el-dataset-en-Training-set-y-Test-set" data-toc-modified-id="Dividir-el-dataset-en-Training-set-y-Test-set-5.1"><span class="toc-item-num">5.1&nbsp;&nbsp;</span>Dividir el dataset en Training set y Test set</a></span></li></ul></li><li><span><a href="#Validacion-y-Evaluacion-Cruzada-(k-fold-Cross-Validation)" data-toc-modified-id="Validacion-y-Evaluacion-Cruzada-(k-fold-Cross-Validation)-6"><span class="toc-item-num">6&nbsp;&nbsp;</span>Validacion y Evaluacion Cruzada (k-fold Cross Validation)</a></span></li><li><span><a href="#Optimizacion-de-Hiper-parametros-(Hyper-Parameter-optimization)" data-toc-modified-id="Optimizacion-de-Hiper-parametros-(Hyper-Parameter-optimization)-7"><span class="toc-item-num">7&nbsp;&nbsp;</span>Optimizacion de Hiper parametros (Hyper Parameter optimization)</a></span></li><li><span><a href="#Evaluacion-final-del-modelo-con-el-Test-set" data-toc-modified-id="Evaluacion-final-del-modelo-con-el-Test-set-8"><span class="toc-item-num">8&nbsp;&nbsp;</span>Evaluacion final del modelo con el Test set</a></span></li><li><span><a href="#Implementacion-del-Modelo-(Deploying)" data-toc-modified-id="Implementacion-del-Modelo-(Deploying)-9"><span class="toc-item-num">9&nbsp;&nbsp;</span>Implementacion del Modelo (Deploying)</a></span></li><li><span><a href="#Comunicacion-de-Resultados-(Data-Story-Telling)" data-toc-modified-id="Comunicacion-de-Resultados-(Data-Story-Telling)-10"><span class="toc-item-num">10&nbsp;&nbsp;</span>Comunicacion de Resultados (Data Story Telling)</a></span></li><li><span><a href="#Conclusiones" data-toc-modified-id="Conclusiones-11"><span class="toc-item-num">11&nbsp;&nbsp;</span>Conclusiones</a></span></li><li><span><a href="#Ayudas-Y-Referencias" data-toc-modified-id="Ayudas-Y-Referencias-12"><span class="toc-item-num">12&nbsp;&nbsp;</span>Ayudas Y Referencias</a></span></li></ul></div>

## Objetivo del Trabajo <a class="tocSkip">
El objetivo inicial del trabajo es utilizar Python para el procesamiento, descripcion  y visualización de datos, con el fin de dejar los datos preparados para ser usados con algoritmos de Machine Learning para Regresión o Clasificación como objetivo final del trabajo.

El trabajo se realizara en Python usando un jupyter notebook o un Notebook de Google Collaboratory, llenando los campos de este archivo y subirlo a github.

## Las actividades a realizar<a class="tocSkip">
    
    
1) Preparar los datos, hacer cada uno de los pasos solo si es necesario:
      
    - Poner los datos en sus tipos de datos correctos
    - Remover los valores duplicados
    - Procesar y Reemplazar los datos que faltan (Missing N.A values)
    - Manejo de outliers
    - Re-escalar las variables
    - Normalizar
    - Data Binning
        - Conversión Numérico a categórico
        - Conversión Categórico a Numérico

2) Realizar una exploración estadística y con visualización de los datos
    - Scatter plots
    - Histogramas
    - boxplot, o las gráficas que considere necesarias


3) Entrene y Evalue modelos de Machine Learning según sea su problema, recuerde usar k-fold cross-validation para evitar over fitting, algunos algoritmos recomendados

    (a) PREDICCIÓN:
    
            ▪ Regresión Lineal
            ▪ Lasso
            ▪ Ridge
            ▪ Decision Tree regressor
            ▪ Random Forest regressor
            ▪ SVR
            ▪ KNR
            ▪ catboost regressor
            
    
    
    (b) CLASIFICACION
      
            ▪ Regresión Logística
            ▪ LinearSVC
            ▪ KernelSVC
            ▪ Decision Tree clasifier
            ▪ Random Forest classifier
            ▪ K-NN
            ▪ GaussianNB
            ▪ catboost
    
**Nota: utilizar visualizaciones para comunicar los resultados obtenidos.**    
            
4) Realizar hyperparameter optimization  de los dos métodos que tengan mejor resultado en los pasos anteriores para mejorar el desempeño obtenido. (Tenga en cuenta, que realizar el hyperparameter tuning no garantiza una mejora significativa en el desempeño, pero es bueno intentarlo)
    
5) Realizar comentarios de cada paso que va realizando en el documento y finalmente también agregue **conclusiones** de:
    
    - los datos procesados
    - la información que observa
    - Comparación de la evaluación de los Métodos
    - Cual fue el mejor modelo que obtuvo luego de la hiper parametrización
    
    
*NOTA: No dude en contactarme para cualquier pregunta o inquietud :)

## FORMATO DE ENTREGA  <a class="tocSkip">

EL jupyter notebook estara alojado en un repositorio de Github

El código debe tener comentarios y explicaciones de la solución del trabajo.


## EVALUACION <a class="tocSkip">
(33 % ) preparacion, descripcion y visualizacion de datos
    
(33 % ) Machine Learning, hiperparametrizacion y Conclusiones
    

|Porcentaje en la evaluación | Descripción| Nada | Incompleto | Completo |
| :---: |:---: |:---: |:---: |:---: |
| 11 % |**Preparación de los datos** <br> (datos nulos, outliers, duplicados, <br> corrección de formatos de tipos de datos) |
|11 % | **Visualización de datos** <br> (Univarible y Bivariable) <br> Histogramas, boxplots, correlaciones, etc|
|11 % | **Descripción y Análisis Estadístico de los datos** |
|11 % | **Machine Learning** <br> Entrenar y evaluar todos los modelos propuestos |
|11 % | **Hiper parametrizacion** <br> Hiperparametrizar 2 modelos y escoger el mejor modelo |   
|11 % | **Resultados y Conclusiones** <br> analisis de los datos, conclusiones del modelos y los resultados obtenidos|


Ejemplos y links de ayuda al final del documento

# Definir el Problema a Resolver

Dado un dataset basado en los precios de las casas en Bostón, se busca determinar el valor de la misma de acuerdo a diferentes características que se almacenan en el dataset como lo son la tasa de criminalidad, la proporción de zona residencial en viviendas de más de 25.000 sq.ff. y entre otras variables que permitirán aproximar el valor de cada vivienda. 

## Describir los datos de entrada y salida

Se cuenta con un total de 17 variables en el dataset.

**Variables de entrada**

Variable | Descripción  | 
 :---:   | :---: | 
zn       | Proporción de terrenos residenciales para lotes de más de 25.000 sq.ff.
black    | 1000 (Bk - 0,63) 2 donde Bk es la proporción de negros por ciudad. 
rad      | Índice de accesibilidad a las autopistas radiales.
indus    | Proporción de hectáreas comerciales no minoristas por ciudad.
ptratio  | Proporción alumnos-profesores por ciudad.
lstat    | Menor estatus de la población (en porcentaje).
nbgde    | **No se tiene concocimiento de esta variable**
chas     | Variable ficticia del río Charles (1 si el tramo limita con el río, 0 en caso contrario).
ftyrgv   | **No se tiene concocimiento de esta variable**
ID       | Identificador de cada casa.
crim     | Tasa de delincuencia per cápita por ciudad.
dis      | Media ponderada de las distancias a cinco centros de empleo de Boston.
tax      | Tasa total de impuesto a la propiedad por 10.000 dólares.
rm       | Número promedio de habitaciones por vivienda.
age      | Proporción de unidades ocupadas por sus propietarios construidas antes de 1940.
nox      | Concentración de óxidos de nitrógeno (partes por cada 10 millones).

**Variable de salida**

Variable | Descripción  | 
 :---:   | :---: | 
medv     | Valor promedio de las viviendas ocupadas por sus propietarios en $1000.

# Importar Librerias

In [1]:
import numpy as np
import pandas as pd
import plotly.express as px

from sklearn.model_selection import train_test_split
from pandas_profiling import ProfileReport

pd.options.plotting.backend = "plotly"

# Cargar Datasets

In [2]:
data_raw = pd.read_csv("../data/raw/Precios_Casas_BostonDS.csv")

# Descripcion General del Dataset
- numero de filas y columnas
- tipos de datos y si estan correctos

In [3]:
data_raw.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 13440 entries, 0 to 13439
Data columns (total 19 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   Unnamed: 0  13440 non-null  int64 
 1   index       13440 non-null  int64 
 2   zn          12925 non-null  object
 3   black       12895 non-null  object
 4   rad         12895 non-null  object
 5   indus       12850 non-null  object
 6   ptratio     12980 non-null  object
 7   medv        12675 non-null  object
 8   lstat       12970 non-null  object
 9   nbgde       12970 non-null  object
 10  chas        13010 non-null  object
 11  ftyrgv      13440 non-null  int64 
 12  ID          12995 non-null  object
 13  crim        12890 non-null  object
 14  dis         12820 non-null  object
 15  tax         12835 non-null  object
 16  rm          12900 non-null  object
 17  age         12890 non-null  object
 18  nox         13000 non-null  object
dtypes: int64(3), object(16)
memory usage: 1.9+ MB


Cantidad de datos inicial:
- 13440 filas, 17 Columns , memory usage: 1.9+ MB

# Limpieza de calidad de datos general
- Filas con valores exactamente iguales (duplicados)
- Columnas duplicadas
- Columnas con valores constantes o sin informacion

### Se verifica existencia de valores duplicados

In [4]:
data_raw.duplicated()

0        False
1        False
2        False
3        False
4        False
         ...  
13435    False
13436    False
13437    False
13438    False
13439    False
Length: 13440, dtype: bool

### Eliminación de columnas innecesarias 

Se eliminan variables existentes en el dataset que no brindan información relevante y se desconoce la relación directa con la variable de salida (medv)

In [5]:
data_raw.drop(['ftyrgv','Unnamed: 0', 'nbgde','ID','index'], axis= 1, inplace = True)

In [6]:
data_raw.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 13440 entries, 0 to 13439
Data columns (total 14 columns):
 #   Column   Non-Null Count  Dtype 
---  ------   --------------  ----- 
 0   zn       12925 non-null  object
 1   black    12895 non-null  object
 2   rad      12895 non-null  object
 3   indus    12850 non-null  object
 4   ptratio  12980 non-null  object
 5   medv     12675 non-null  object
 6   lstat    12970 non-null  object
 7   chas     13010 non-null  object
 8   crim     12890 non-null  object
 9   dis      12820 non-null  object
 10  tax      12835 non-null  object
 11  rm       12900 non-null  object
 12  age      12890 non-null  object
 13  nox      13000 non-null  object
dtypes: object(14)
memory usage: 1.4+ MB


### Reemplazo de valores faltantes
- Para más facilidad se decide reemplazar los valores faltantes (NaN) dentro del data set por el caracter '?' ya que esto nos permite dividir los datos en el siguiente paso.

In [7]:
data_raw.replace(np.nan, '?', inplace = True)

Se verifica la existencia de valores nulos en el dataset

In [8]:
data_raw.isna()

zn  black    rad  indus  ptratio   medv  lstat   chas   crim    dis  \
0      False  False  False  False    False  False  False  False  False  False   
1      False  False  False  False    False  False  False  False  False  False   
2      False  False  False  False    False  False  False  False  False  False   
3      False  False  False  False    False  False  False  False  False  False   
4      False  False  False  False    False  False  False  False  False  False   
...      ...    ...    ...    ...      ...    ...    ...    ...    ...    ...   
13435  False  False  False  False    False  False  False  False  False  False   
13436  False  False  False  False    False  False  False  False  False  False   
13437  False  False  False  False    False  False  False  False  False  False   
13438  False  False  False  False    False  False  False  False  False  False   
13439  False  False  False  False    False  False  False  False  False  False   

         tax     rm    age    nox  
0      False  False  False  False  
1      False  False  False  False  
2      False  False  False  False  
3      False  False  False  False  
4      False  False  False  False  
...      ...    ...    ...    ...  
13435  False  False  False  False  
13436  False  False  False  False  
13437  False  False  False  False  
13438  False  False  False  False  
13439  False  False  False  False  

[13440 rows x 14 columns]

## Dividir el dataset en Training set y Test set

Este paso debe ser al inicio de los proyectos,  Se deben realizar todas las transformaciones, preparación de datos y limpieza de los datos, en el train set y en la evaluacion se deben aplicarl al test set y a los datos nuevos que lleguen al sistema. Esta division inicial se hace para evitar data leakage de los datos de test a los datos de train, por ejemplo en las imputaciones.

Por este motivo se realizara en esta parte.

Division de los datos de entrenamiento (Train set) y de Evaluacion (test - set), las divisiones utilizadas son

- 80% (train) , 20%(test)
- 70% (train) , 30%(test)

El **Train set** se hace para seleccion de Modelos y el **Test-set** solamente para la evaluacion Final

In [9]:
x = data_raw.drop('medv',axis=1)
y = data_raw['medv']

X_train, X_test, y_train, y_test = train_test_split(x, y, test_size=0.2,random_state = 123, stratify=y)

In [10]:
print('Tamaño del set de entrenamiento', X_train.shape)

Tamaño del set de entrenamiento (10752, 13)


In [11]:
train = pd.concat([X_train,y_train],axis=1)

# Descripcion  y Limpieza de los datos

Se describen nuevamente los datos seleccionados para el entrenamiento (80%).

In [12]:
train.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 10752 entries, 11268 to 8508
Data columns (total 14 columns):
 #   Column   Non-Null Count  Dtype 
---  ------   --------------  ----- 
 0   zn       10752 non-null  object
 1   black    10752 non-null  object
 2   rad      10752 non-null  object
 3   indus    10752 non-null  object
 4   ptratio  10752 non-null  object
 5   lstat    10752 non-null  object
 6   chas     10752 non-null  object
 7   crim     10752 non-null  object
 8   dis      10752 non-null  object
 9   tax      10752 non-null  object
 10  rm       10752 non-null  object
 11  age      10752 non-null  object
 12  nox      10752 non-null  object
 13  medv     10752 non-null  object
dtypes: object(14)
memory usage: 1.2+ MB


### Se obtienen los valores únicos por columna

In [13]:
for col in train:
    print(f"{col} = {train[col].unique()}")

zn = ['gdhg7u8whui 784gryb' '0.0' '?' '33.0' 'xxxxx' '22.0' '20.0' '25.0'
 '-26543765432345.0' '21.0' '90.0' '82.5' '30.0' '45.0' '88876599956788.0'
 '12.5' '95.0' '75.0' '55.0' '80.0' '18.0' '35.0' '34.0' '52.5' '40.0'
 '85.0' '60.0' '28.0' '17.5' '100.0']
black = ['gdhg7u8whui 784gryb' '385.91' '262.76' '393.36' '341.6' '378.38'
 '390.18' '393.68' '396.9' '48.45' '390.39' '353.89' '390.77' '?' '372.8'
 '378.08' '240.52' '394.87' '88876599956788.0' '7.68' '350.45' '395.56'
 '-26543765432345.0' '386.85' '394.72' '392.63' '378.25' '347.88' '81.33'
 '383.32' '255.23' '60.72' '392.85' '394.96' '392.11' 'xxxxx' '391.25'
 '377.68' '334.4' '387.89' '395.43' '395.52' '390.64' '396.06' '321.02'
 '352.58' '370.73' '388.52' '376.73' '359.29' '386.73' '394.51' '261.95'
 '366.15' '88.63' '285.83' '330.04' '210.97' '9.32' '3.65' '395.6'
 '374.43' '395.62' '388.45' '392.68' '390.55' '391.5' '100.63' '375.87'
 '391.83' '377.67' '356.99' '314.64' '396.21' '394.08' '383.73' '100.19'
 '22.01' '378.95' '

### Análisis de resultados
- Todas las columnas contienen los valores ['xxxxx', 'gdhg7u8whui 784gryb', '-26543765432345.0', '88876599956788.0', '?'] que configuran un ruido para el entrenamiento.

Se reemplazan los valores desconocidos por el valor NaN para luego poderlos eliminar mediante la función 'dropna'

In [14]:
train.replace(['xxxxx', '-26543765432345.0', 'gdhg7u8whui 784gryb', '88876599956788.0'], np.nan, inplace = True)

In [15]:
train.dropna(inplace=True)

### Restablecer los valores nulos del dataset de entrenamiento
Se recuperan los valores NaN que anteriormente se habían cambiado por el caracter '?' en el paso de particionamiento del dataset.

In [16]:
train.replace('?', np.nan, inplace = True)

In [17]:
train.reset_index(inplace = True, drop = True)

In [18]:
train.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3950 entries, 0 to 3949
Data columns (total 14 columns):
 #   Column   Non-Null Count  Dtype 
---  ------   --------------  ----- 
 0   zn       3563 non-null   object
 1   black    3565 non-null   object
 2   rad      3584 non-null   object
 3   indus    3565 non-null   object
 4   ptratio  3612 non-null   object
 5   lstat    3599 non-null   object
 6   chas     3641 non-null   object
 7   crim     3562 non-null   object
 8   dis      3524 non-null   object
 9   tax      3525 non-null   object
 10  rm       3595 non-null   object
 11  age      3591 non-null   object
 12  nox      3650 non-null   object
 13  medv     3537 non-null   object
dtypes: object(14)
memory usage: 432.2+ KB


Análisis previo del set de entrenamiento: 
- Filas: 4048
- Columnas: 14
- Memoria 432.2+ KB

## Perfilamiento del set de entrenamiento - reporte 1

**Nota:** recordar que los datos NaN en el dataset original se cambiaron por el caracter '?' para poder realizar el particionamiento.

In [19]:
profile = ProfileReport(train, title="Boston House Prices Report 1",
                        explorative=True, samples=None, correlations=None, 
                        missing_diagrams=None, duplicates=None, interactions=None)

In [20]:
profile.to_notebook_iframe()

Summarize dataset:   0%|          | 0/23 [00:00<?, ?it/s]

Generate report structure:   0%|          | 0/1 [00:00<?, ?it/s]

Render HTML:   0%|          | 0/1 [00:00<?, ?it/s]

In [21]:
profile.to_file("../reports/Boston_house_prices_report_1.html")

Export report to file:   0%|          | 0/1 [00:00<?, ?it/s]

#### Análisis de perfil generado

Variable | Análisis  | 
 :---:   | :---: | 
zn       | Para el 69.6% de datos la proporción de terrenos residenciales para lotes de más de 25.000 m² es igual a 0.
black    | Variable con ALTA CARDINALIDAD, 237 valores diferentes, 396.9 20.5% más común, el 55.6% de los valores se reconoce como (other).
rad      | Variable balanceada, 9% de valores faltantes, 24.0 25.6% más común.
indus    | Variable con ALTA CARDINALIDAD, 68 valores diferentes, 18.1 25.8% más común, el 15.9% de los valores se reconoce como (other).
ptratio  | Variable con el 8.3% de valores faltantes, 20.2 26.5% más común.
lstat    | Variable con ALTA CARDINALIDAD, 310 valores diferentes, el 74% de los valores se reconoce como (other).
chas     | Variable desbalanceada, 7.6% de valores faltantes, el 86.9% de casas no limita con el río.
crim     | Variable con ALTA CARDINALIDAD, 332 valores diferentes, 9.6% de valores faltantes, el 75.9% se reconoce como (other).
dis      | Variable con ALTA CARDINALIDAD, 295 valores diferentes, 10.5% de valores faltantes, el 72.4% se reconoce como (other).
tax      | Variable con ALTA CARDINALIDAD, 59 valores diferentes, 666.0 25.4% más común. 
rm       | Variable con ALTA CARDINALIDAD, 308 valores diferentes, 8.8% de valores faltantes, el 70% de los valores se reconoce como (other).
age      | Variable con ALTA CARDINALIDAD, 260 valores diferentes, 100 8.8% más común, el 63.2% de los valores se reconoce como (other).
nox      | Variable con ALTA CARDINALIDAD, 77 valores diferentes, 605 4.9% más común, 7.4% de valores faltantes.

**Warnings** <br>

El dataset tiene 3280(81%) de filas duplicadas.

### Cantidad de valores nulos existentes en el set de entrenamiento

In [22]:
(train.isna().sum()*100/len(train)).plot.bar(title = 'Percentage of Missing Values per column',color=train.columns)

### Porcentaje de valores nulos en el set de entrenamiento

In [23]:
missing = train.isna().sum()
missing[missing>0]*100/len(train)

zn          9.797468
black       9.746835
rad         9.265823
indus       9.746835
ptratio     8.556962
lstat       8.886076
chas        7.822785
crim        9.822785
dis        10.784810
tax        10.759494
rm          8.987342
age         9.088608
nox         7.594937
medv       10.455696
dtype: float64

## Analisis General Univariable y Bivariable 
Analisis de cada una de las variables para lograr calidad de datos en cada columna
- **Correccion del tipo de dato (numericas, categoricas, string) de cada columna (optimizar memoria)**
- Deteccion de numero de datos faltantes
- Deteccion de duplicados

### Limpieza de los datos

- Es necesario hacer un dataset auxiliar para allí eliminarlos temporalmente y poder obtener las medidas de tendencia central (moda, mediana y media) necesarias para la imputación de datos faltantes.

In [24]:
def columnTreatment(column, data):
    dataColumn = data[column].dropna()
    dataColumn = dataColumn.astype('float64')

    dataFrame = pd.DataFrame(dataColumn)
    dataFrame.reset_index(inplace = True, drop = True)

    return dataFrame

In [25]:
def getQuartiles(dataFrame, column):
    Q1 = np.percentile(dataFrame[column].dropna(), 25, interpolation = 'midpoint')
    Q3 = np.percentile(dataFrame[column].dropna(), 75, interpolation = 'midpoint')

    IQR = Q3 - Q1

    return {'Q1': Q1,'Q3': Q3, 'IQR': IQR}

In [26]:
def getBounds(dataFrame, column, quartiles):
    Q1 = quartiles['Q1']
    Q3 = quartiles['Q3']
    IQR = quartiles['IQR']

    print('Q1', Q1)
    print('Q3', Q3)
    print('IQR', IQR)

    # Upper bound
    upper = np.where(dataFrame[column] >= (Q3 + 1.5*IQR))

    # Below Lower bound
    lower = np.where(dataFrame[column] <= (Q1 - 1.5*IQR))

    return {'upper': upper, 'lower': lower}

In [27]:
def deleteOutliersDataframe(dataFrame, bounds):
    upper = bounds['upper']
    lower = bounds['lower']

    dataFrame.drop(upper[0], inplace = True)
    dataFrame.drop(lower[0], inplace = True)
    dataFrame.reset_index(inplace = True, drop = True)

    return dataFrame

In [28]:
def deleteOutliersSetTrain(train, column):
    quartiles = getQuartiles(train.copy(), column)
    bounds = getBounds(train.copy(), column, quartiles)
    train = deleteOutliersDataframe(train.copy(), bounds)

    return train

In [29]:
def generateDataframe(train, column):
    dataFrame = columnTreatment(column, train.copy())
    quartiles = getQuartiles(dataFrame.copy(), column)
    bounds = getBounds(dataFrame.copy(), column, quartiles)

    dataFrame = deleteOutliersDataframe(dataFrame.copy(), bounds)

    return dataFrame

### Análisis varible CRIM

In [30]:
train['crim'].isna().sum()

388

Se evidencia que la variable "crim" contiene 388 datos nulos, asi que vamos a proceder a ver la distribucion de los datos para analizar que se puede hacer con los datos faltantes

In [31]:
crim = train['crim'].dropna().astype('float64')
px.box(crim,title = "crim")

In [32]:
crim.describe()

count     3562.000000
mean       112.790874
std       1134.295032
min          0.006320
25%          0.093780
50%          0.382140
75%          5.090170
max      15288.000000
Name: crim, dtype: float64

La distribucion de los datos se ve muy contaminada por los valores atipicos entonces vamos a proceder a eliminarlos

In [33]:
dfCrim = generateDataframe(train.copy(), 'crim')

Q1 0.09378
Q3 5.09017
IQR 4.99639
Upper bound: (array([   5,    6,   24,   26,   31,   32,   47,   56,   67,   79,   98,
        112,  113,  120,  139,  149,  158,  159,  190,  193,  208,  209,
        211,  222,  244,  246,  254,  260,  276,  281,  285,  287,  294,
        311,  315,  332,  347,  350,  357,  367,  372,  377,  382,  397,
        407,  414,  429,  434,  441,  461,  467,  477,  482,  489,  494,
        508,  517,  543,  552,  555,  560,  566,  568,  572,  581,  588,
        592,  621,  624,  625,  638,  646,  647,  677,  684,  688,  695,
        705,  716,  730,  732,  749,  750,  755,  779,  787,  796,  797,
        812,  813,  828,  849,  864,  873,  875,  899,  917,  940,  944,
        946,  967,  992, 1009, 1011, 1013, 1015, 1017, 1035, 1048, 1051,
       1054, 1059, 1060, 1071, 1084, 1091, 1092, 1102, 1111, 1112, 1118,
       1141, 1147, 1170, 1178, 1188, 1194, 1198, 1202, 1205, 1210, 1214,
       1223, 1228, 1229, 1230, 1231, 1257, 1269, 1277, 1279, 1283, 1288,
   

In [34]:
px.box(dfCrim,title = "crim")

A este punto aun conservamos algunos valores atipicos, pero a diferencia de la versión inicial esta ya esta mas depurada

In [35]:
meanCrim = dfCrim['crim'].mean()
print(meanCrim)

1.9072951445631734


In [36]:
px.histogram(dfCrim,x="crim")

Rellenamos los datos que estan como nulos por el valor de la media esto según se ve en la histograma no altera la distribución de forma significativa

In [37]:
train['crim'].replace(np.nan,meanCrim,inplace=True)
train['crim'] = train['crim'].astype('float64')
train['crim'].describe()

count     3950.000000
mean       101.899018
std       1077.636376
min          0.006320
25%          0.105740
50%          0.761620
75%          4.261310
max      15288.000000
Name: crim, dtype: float64

Luego de la imputación debemos recordar que los datos atipicos se quitaron para encontrar la media, entonces en esta ocacion debemos quitar los datos atipicos de nuestro set de entrenamiento

In [38]:
train.reset_index(inplace = True, drop = True)
train = deleteOutliersSetTrain(train,'crim')
train['crim'] = train['crim'].astype('float64')
train['crim'].describe() 

Q1 0.10574
Q3 4.26131
IQR 4.15557
Upper bound: (array([   6,    7,   27,   29,   35,   36,   47,   52,   61,   74,   87,
        108,  124,  125,  133,  139,  154,  165,  174,  175,  200,  208,
        211,  227,  228,  230,  241,  264,  266,  274,  280,  299,  305,
        310,  312,  320,  339,  344,  362,  369,  378,  381,  388,  398,
        404,  410,  417,  434,  446,  454,  456,  457,  472,  478,  487,
        510,  516,  528,  534,  541,  547,  558,  561,  570,  598,  604,
        608,  611,  616,  623,  625,  629,  639,  647,  651,  684,  687,
        688,  701,  709,  711,  744,  752,  757,  764,  776,  778,  789,
        806,  808,  828,  829,  834,  860,  869,  880,  881,  896,  897,
        913,  935,  949,  952,  964,  966,  993, 1004, 1011, 1038, 1042,
       1044, 1050, 1051, 1068, 1097, 1114, 1116, 1118, 1120, 1122, 1143,
       1156, 1159, 1162, 1167, 1168, 1181, 1195, 1203, 1204, 1208, 1214,
       1217, 1220, 1223, 1224, 1232, 1258, 1265, 1266, 1289, 1299, 1309,
   

count    3514.000000
mean        1.753721
std         2.586940
min         0.006320
25%         0.092990
50%         0.352330
75%         1.907295
max        10.062300
Name: crim, dtype: float64

In [39]:
px.histogram(train['crim'],x="crim")

### Analisis variable ZN

In [40]:
zn = train['zn'].dropna().astype('float64')
px.box(zn,title = "zn")

In [41]:
px.histogram(zn,x="zn")

In [42]:
train['zn'].describe()
print(train['zn'].value_counts())

0.0      2351
20.0      158
12.5       78
22.0       59
75.0       52
33.0       45
80.0       45
34.0       44
25.0       43
45.0       34
95.0       33
85.0       26
55.0       26
90.0       25
60.0       23
28.0       22
21.0       20
30.0       20
52.5       14
40.0       14
17.5        9
82.5        7
100.0       7
35.0        7
18.0        5
Name: zn, dtype: int64


In [43]:
train['zn'] = train['zn'].astype('float64')

In [44]:
train['zn'].value_counts()[0]*100/len(train['zn'])

66.90381331815595

Teniendo en cuenta que el 66 % de los datos son ceros los datos que estan como null se van a llenar con los datos de la moda en este caso 0

In [45]:
train['zn'].fillna(train['zn'].mode()[0], inplace=True)

In [46]:
train['zn'].isna().sum()

0

### Analisis variable INDUS

In [47]:
indus = train['indus'].dropna().astype('float64')
indus.describe()

count    3157.000000
mean       11.537555
std         7.187196
min         0.740000
25%         5.190000
50%         9.900000
75%        18.100000
max        27.740000
Name: indus, dtype: float64

In [48]:
px.box(indus,title = "indus")

Los datos par este caso no tiene outliers

In [49]:
px.histogram(indus,x="indus")

In [50]:
train['indus'].isna().sum()

357

Por como esta la distribución vamos a descargar los datos nullos ya que para hacer una imputacion se deberian tener muchas factores en cuenta

In [51]:
train.drop(train[train['indus'].notnull() == False].index, inplace=True)

In [52]:
train['indus'] = train['indus'].astype('float64')

In [53]:
train['indus'].describe()

count    3157.000000
mean       11.537555
std         7.187196
min         0.740000
25%         5.190000
50%         9.900000
75%        18.100000
max        27.740000
Name: indus, dtype: float64

In [54]:
train['indus'].isna().sum()

0

### Analisis variable CHAS

In [55]:
chas = train['chas'].dropna().astype('float64')
print(train['chas'].value_counts())
px.histogram(chas,x="chas")

0.0    2764
1.0     188
Name: chas, dtype: int64


In [56]:
train['chas'] = train['chas'].astype('float64')

In [57]:
train['chas'].value_counts()

0.0    2764
1.0     188
Name: chas, dtype: int64

In [58]:
train['chas'].dropna().sum()

188.0

Teniendo en cuenta esta distribución vamos a imputar los datos faltantes como 0.0

In [59]:
train['chas'].fillna(train['chas'].mode()[0], inplace=True)

In [60]:
train['chas'].value_counts()

0.0    2969
1.0     188
Name: chas, dtype: int64

### Analisis variable NOX

In [61]:
noxx = train['nox'].dropna().astype('float64')
noxx.describe()

count    2952.000000
mean      479.829489
std       221.523143
min         0.400000
25%       433.000000
50%       521.000000
75%       605.000000
max       871.000000
Name: nox, dtype: float64

In [62]:
train['nox'].value_counts()

605.0    162
871.0    155
713.0    147
538.0    129
624.0    119
        ... 
389.0      4
0.55       4
447.0      4
404.0      3
469.0      3
Name: nox, Length: 76, dtype: int64

In [63]:
px.box(noxx,title = "nox")

In [64]:
px.histogram(noxx,x="nox")

In [65]:
train.reset_index(inplace = True, drop = True)
train['nox'] = train['nox'].astype('float64')
train = deleteOutliersSetTrain(train,'nox')
train['nox'].describe()

Q1 433.0
Q3 605.0
IQR 172.0
Upper bound: (array([  38,   43,   51,   53,   62,   66,   78,   88,  113,  128,  129,
        171,  172,  176,  210,  216,  229,  241,  278,  282,  307,  328,
        355,  457,  502,  505,  562,  579,  675,  686,  748,  783,  819,
        840,  853,  861,  882,  888,  922,  972,  984, 1058, 1062, 1069,
       1101, 1119, 1214, 1225, 1226, 1256, 1262, 1271, 1306, 1310, 1319,
       1335, 1352, 1357, 1359, 1367, 1369, 1372, 1428, 1435, 1454, 1477,
       1563, 1575, 1576, 1583, 1590, 1591, 1596, 1604, 1611, 1621, 1633,
       1635, 1647, 1659, 1691, 1722, 1730, 1751, 1765, 1773, 1807, 1823,
       1882, 1883, 1884, 1917, 1918, 1922, 1928, 1936, 1944, 1995, 2003,
       2041, 2087, 2111, 2132, 2152, 2153, 2178, 2183, 2209, 2226, 2244,
       2254, 2262, 2263, 2271, 2304, 2314, 2321, 2344, 2360, 2364, 2378,
       2391, 2402, 2414, 2430, 2545, 2558, 2560, 2564, 2589, 2687, 2697,
       2705, 2722, 2743, 2746, 2780, 2807, 2812, 2825, 2856, 2862, 2882,
       28

count    2385.000000
mean      537.195388
std        88.377633
min       385.000000
25%       449.000000
50%       538.000000
75%       605.000000
max       718.000000
Name: nox, dtype: float64

In [66]:
px.box(train['nox'],title = "nox")

In [67]:
px.histogram(train['nox'],x="nox")

Par este caso tambien vamos a descartar los valores que estanc como null

In [68]:
train.drop(train[train['nox'].notnull() == False].index, inplace=True)

In [69]:
train['nox'].isna().sum()

0

### Analisis variable RM

In [70]:
rm = train['rm'].dropna().astype('float64')
rm.describe()

count    2180.000000
mean     5590.114321
std      2146.122089
min         4.970000
25%      5834.000000
50%      6122.000000
75%      6549.000000
max      8725.000000
Name: rm, dtype: float64

In [71]:
px.box(rm,title = "rm")

En el Boxplot se evidencia que esta variable tiene valores atipicos entonces se procede a quitarlos

In [72]:
train.reset_index(inplace = True, drop = True)
train['rm'] = train['rm'].astype('float64')
train = deleteOutliersSetTrain(train,'rm')
train['rm'].describe()

Q1 5834.0
Q3 6549.0
IQR 715.0
Upper bound: (array([   1,   12,   22,  107,  113,  124,  183,  204,  218,  223,  230,
        251,  256,  268,  269,  287,  347,  360,  393,  396,  406,  411,
        422,  430,  431,  454,  467,  495,  513,  521,  525,  531,  542,
        554,  569,  627,  636,  669,  676,  681,  696,  706,  708,  717,
        772,  786,  811,  887,  904,  923,  966, 1035, 1053, 1086, 1088,
       1093, 1106, 1135, 1174, 1186, 1194, 1229, 1275, 1279, 1305, 1325,
       1331, 1335, 1338, 1341, 1350, 1354, 1406, 1409, 1468, 1497, 1509,
       1522, 1523, 1536, 1565, 1578, 1624, 1661, 1723, 1736, 1737, 1741,
       1814, 1836, 1845, 1863, 1879, 1944, 1957, 1961, 2003, 2014, 2016,
       2049, 2078, 2079, 2086, 2089, 2096, 2120, 2132, 2134, 2136, 2144,
       2161, 2209, 2230, 2262, 2271, 2283, 2314, 2318, 2343, 2354]),)
Lower bound: (array([  10,   21,   26,   27,   31,   37,   48,   55,   69,   70,   71,
         82,   83,   88,   92,   96,  102,  109,  118,  125,  148,  1

count    1788.000000
mean     6239.605705
std       480.125579
min      4963.000000
25%      5929.750000
50%      6169.000000
75%      6525.000000
max      7489.000000
Name: rm, dtype: float64

In [73]:
px.box(train['rm'],title = "rm")

In [74]:
train['rm'] = train['rm'].astype('float64')
train['rm'].describe()

count    1788.000000
mean     6239.605705
std       480.125579
min      4963.000000
25%      5929.750000
50%      6169.000000
75%      6525.000000
max      7489.000000
Name: rm, dtype: float64

In [75]:
px.histogram(train['rm'],x="rm")

Observando la distribución vamos a descargar los datos que tenemos como null

In [76]:
train.drop(train[train['rm'].notnull() == False].index, inplace=True)

In [77]:
train['rm'].isna().sum()

0

### Análisis varible AGE

In [78]:
age = train['age'].dropna().astype('float64')
age.describe()

count    1687.000000
mean       65.539775
std        28.237261
min         6.000000
25%        41.100000
50%        73.100000
75%        91.400000
max       100.000000
Name: age, dtype: float64

In [79]:
px.box(age,title = "rm")

In [80]:
px.histogram(age,x="age")

In [81]:
train['age'].isna().sum()

101

Observando la distribucion y la cantidad de datos que son que son null vamos a proceder a eliminar los datos null

In [82]:
train.drop(train[train['age'].notnull() == False].index, inplace=True)

In [83]:
train['age'].isna().sum()

0

In [84]:
train.shape

(1687, 14)

### Análisis variable DIS - ALTA CARDINALIDAD

Se verifica la cantidad de valores faltantes en la columna DIS

In [85]:
train['dis'].isna().sum()

85

Se debe generar un dataframe auxiliar para así eliminar temporalmetne los valores NaN y encontrar la media que en este caso nos servirá para imputar los datos faltantes más adelante.

In [86]:
dataframeDis = generateDataframe(train.copy(), 'dis')
meanDis = dataframeDis['dis'].mean()
print('Mean found:' , meanDis)

Q1 2.3682
Q3 5.7209
IQR 3.3527000000000005
Upper bound: (array([   9,   23,   43,   46,   69,   78,   82,   96,  100,  130,  144,
        149,  167,  171,  200,  218,  220,  276,  278,  282,  295,  350,
        351,  367,  369,  371,  385,  391,  408,  414,  418,  431,  436,
        438,  454,  487,  500,  516,  524,  546,  562,  564,  584,  601,
        616,  624,  679,  703,  708,  753,  760,  765,  766,  787,  790,
        812,  836,  841,  858,  859,  897,  898,  905,  915,  922,  961,
        978, 1014, 1020, 1022, 1028, 1032, 1037, 1041, 1052, 1057, 1063,
       1065, 1072, 1095, 1100, 1122, 1129, 1137, 1139, 1140, 1150, 1153,
       1173, 1187, 1188, 1194, 1223, 1225, 1246, 1252, 1254, 1257, 1278,
       1280, 1290, 1295, 1302, 1305, 1316, 1335, 1344, 1351, 1352, 1355,
       1396, 1407, 1444, 1450, 1477, 1516, 1518, 1520, 1558, 1569, 1581,
       1585, 1598]),)
Lower bound: (array([], dtype=int64),)
Mean found: 3.9614782961460517


Columna DIS en el dataset de entrenamiento con valores NaN.

In [87]:
train['dis'].describe()

count       1602
unique       182
top       5.7209
freq          35
Name: dis, dtype: object

Columna DIS en el dataset de entrenamiento con valores imputados dada la media.

In [88]:
train['dis'].fillna(meanDis, inplace = True)
train['dis'] = train['dis'].astype('float64')
train['dis'].describe()

count    1687.000000
mean      289.851661
std      1119.178086
min         1.129600
25%         2.425900
50%         3.945400
75%         5.720900
max      8344.000000
Name: dis, dtype: float64

In [89]:
px.box(train['dis'], title = "BOXPLOT DIAGRAM OF DIS VARIABLE WITH OUTLIERS")

Se borran los outliers en el dataset de entrenamiento.

In [90]:
train.reset_index(inplace = True, drop = True)
train = deleteOutliersSetTrain(train.copy(), 'dis')

Q1 2.4259
Q3 5.7209
IQR 3.2950000000000004
Upper bound: (array([  10,   25,   37,   45,   48,   71,   81,   86,  101,  105,  136,
        150,  155,  176,  180,  210,  228,  229,  231,  264,  290,  292,
        296,  310,  367,  368,  386,  388,  390,  406,  412,  429,  435,
        439,  453,  459,  461,  478,  512,  525,  541,  550,  572,  589,
        591,  612,  631,  646,  655,  686,  712,  738,  743,  791,  798,
        803,  804,  825,  828,  850,  876,  881,  899,  900,  939,  940,
        948,  958,  965, 1004, 1022, 1059, 1066, 1068, 1075, 1079, 1084,
       1089, 1100, 1105, 1113, 1115, 1123, 1148, 1153, 1175, 1182, 1190,
       1192, 1193, 1203, 1206, 1226, 1240, 1241, 1247, 1250, 1279, 1281,
       1306, 1312, 1314, 1317, 1340, 1342, 1353, 1359, 1366, 1369, 1380,
       1399, 1410, 1418, 1419, 1422, 1424, 1464, 1470, 1485, 1522, 1528,
       1555, 1596, 1598, 1600, 1640, 1651, 1664, 1669, 1670, 1683]),)
Lower bound: (array([], dtype=int64),)


In [91]:
px.box(train['dis'], title = "BOXPLOT DIAGRAM OF DIS")

Se verifica la cantidad de valores faltantes para la variable DIS

In [92]:
train['dis'].isna().sum()

0

### Análisis variable RAD

Primero se verifica la cantidad de valores faltantes en la columna rad

In [93]:
train['rad'].isna().sum()

74

Se obtiene el listado de valores únicos que contiene la columna

In [94]:
colum = train['rad'].unique()
colum

array([nan, '8.0', '4.0', '3.0', '6.0', '5.0', '24.0', '2.0', '1.0',
       '7.0'], dtype=object)

Gráfico de barras para observar la distribución de la variable

In [95]:
bar = train['rad'].value_counts()
bar.plot.bar(title="RAD DISTRIBUTION", color=bar.index)

El mayor índice de accesibilidad a las autopistas es 24.0, así que se imputan los datos faltantes tomando la moda.

In [96]:
train['rad'].fillna(train['rad'].mode()[0], inplace = True)

Se verifica la cantidad de valores faltantes para la variable RAD

In [97]:
train['rad'].isna().sum()

0

### Análisis variable TAX - ALTA CARDINALIDAD

Primero se verifica la cantidad de valores faltantes en la columna tax.

In [98]:
train['tax'].isna().sum()

102

Gráfico de barras para observar la distribución de la variable

In [99]:
bar = train['tax'].value_counts()
bar.plot.bar(title="TAX DISTRIBUTION", color=bar.index)

In [100]:
px.box(train['tax'], title = "BOXPLOT DIAGRAM OF TAX VARIABLE")

El valor que más se repite en la columna TAX es 666.0, así que se imputan los datos faltantes tomando la moda.

In [101]:
train['tax'].fillna(train['tax'].mode()[0], inplace = True)

Se verifica la cantidad de valores faltantes para la variable TAX.

In [102]:
train['tax'].isna().sum()

0

### Análisis variable PTRATIO

Primero se verifica la cantidad de valores faltantes en la columna ptratio.

In [103]:
train['ptratio'].isna().sum()

48

Se obtiene el listado de valores únicos que contiene la columna.

In [104]:
colum = train['ptratio'].unique()
colum

array(['19.1', '19.7', '21.2', '18.6', '16.6', '15.2', '19.6', '20.2',
       '13.0', nan, '14.7', '17.9', '16.8', '18.4', '15.3', '16.1',
       '19.2', '17.8', '21.0', '20.1', '16.0', '15.1', '17.4', '18.2',
       '18.3', '18.5', '15.6', '19.0', '18.7', '17.0', '18.0', '17.6',
       '15.5'], dtype=object)

Gráfico de barras para observar la distribución de la variable.

In [105]:
bar = train['ptratio'].value_counts()
bar.plot.bar(title="PTRATIO DISTRIBUTION", color=bar.index)

In [106]:
px.box(train['ptratio'], title = "BOXPLOT DIAGRAM OF PTRATIO VARIABLE")

El valor que más se repite en la columna PTRATIO es 20.2, así que se imputan los datos faltantes tomando la moda.

In [107]:
train['ptratio'].fillna(train['ptratio'].mode()[0], inplace = True)

Se verifica la cantidad de valores faltantes para la variable PTRATIO

In [108]:
train['ptratio'].isna().sum()

0

### Análisis variable BLACK - ALTA CARDINALIDAD

Primero se verifica la cantidad de valores faltantes en la columna black.

In [109]:
train['black'].isna().sum()

91

Gráfico de barras para observar la distribución de la variable.

In [110]:
bar = train['black'].value_counts()
bar.plot.bar(title="BLACK DISTRIBUTION", color=bar.index)

In [111]:
px.box(train['black'], title = "BOXPLOT DIAGRAM OF BLACK VARIABLE WITH OUTLIERS")

Se debe generar un dataframe auxiliar para así eliminar temporalmente los valores NaN y encontrar la moda que en este caso nos servirá para imputar los datos faltantes más adelante.

In [112]:
dataframeBlack = generateDataframe(train.copy(), 'black')
modeBlack = dataframeBlack['black'].mode()[0]
print('Mode found:' , modeBlack)

Q1 374.43
Q3 396.21
IQR 21.779999999999973
Upper bound: (array([], dtype=int64),)
Lower bound: (array([   6,   13,   14,   22,   27,   39,   42,   46,   51,   57,   67,
         70,   79,   80,   89,   91,  102,  110,  113,  118,  122,  123,
        126,  131,  134,  135,  137,  143,  144,  149,  161,  163,  164,
        172,  182,  193,  200,  214,  228,  235,  238,  245,  250,  251,
        264,  270,  284,  289,  300,  304,  306,  307,  313,  316,  320,
        322,  328,  333,  337,  341,  350,  357,  360,  372,  382,  399,
        403,  407,  409,  414,  418,  421,  437,  439,  445,  446,  462,
        463,  474,  475,  478,  481,  484,  485,  507,  509,  511,  514,
        526,  535,  541,  550,  555,  562,  564,  567,  576,  579,  587,
        591,  609,  614,  616,  619,  624,  627,  628,  629,  648,  649,
        651,  657,  663,  665,  669,  671,  678,  679,  685,  688,  689,
        691,  692,  697,  700,  701,  708,  713,  718,  721,  725,  727,
        728,  729,  735,  74

Columna BLACK en el dataset de entrenamiento con valores NaN.

In [113]:
train['black'].describe()

count      1465
unique      141
top       396.9
freq        342
Name: black, dtype: object

Columna BLACK en el dataset de entrenamiento con valores imputados dada la moda.

In [114]:
train['black'].fillna(modeBlack, inplace = True)
train['black'] = train['black'].astype('float64')
train['black'].describe()

count    1556.000000
mean      360.167352
std        81.917837
min         3.500000
25%       374.710000
50%       392.530000
75%       396.900000
max       396.900000
Name: black, dtype: float64

Se borran los outliers en el dataset de entrenamiento.

In [115]:
train.reset_index(inplace = True, drop = True)
train = deleteOutliersSetTrain(train.copy(), 'black')

Q1 374.71
Q3 396.9
IQR 22.189999999999998
Upper bound: (array([], dtype=int64),)
Lower bound: (array([   7,   16,   17,   25,   30,   42,   45,   50,   56,   62,   72,
         75,   84,   85,   95,   97,  108,  118,  121,  126,  131,  132,
        135,  141,  144,  145,  147,  154,  155,  161,  173,  175,  176,
        184,  194,  205,  213,  228,  243,  251,  255,  262,  268,  269,
        284,  290,  304,  309,  321,  326,  328,  329,  335,  338,  343,
        345,  351,  356,  360,  364,  374,  381,  384,  396,  408,  427,
        431,  435,  437,  443,  448,  451,  468,  470,  477,  478,  496,
        497,  508,  509,  512,  515,  518,  519,  542,  544,  546,  549,
        561,  570,  576,  585,  591,  598,  600,  603,  612,  615,  626,
        630,  649,  654,  657,  660,  665,  669,  670,  671,  690,  691,
        693,  699,  706,  708,  712,  714,  722,  723,  730,  733,  734,
        736,  737,  742,  746,  747,  754,  760,  765,  768,  772,  774,
        775,  776,  782,  791

Se verifica la cantidad de valores faltantes para la variable BLACK.

In [116]:
train['black'].isna().sum()

0

### Análisis variable LSTAT - ALTA CARDINALIDAD

Primero se verifica la cantidad de valores faltantes en la columna LSTAT.

In [117]:
train['lstat'].isna().sum()

75

Gráfico de barras para observar la distribución de la variable

In [118]:
bar = train['lstat'].value_counts()
bar.plot.bar(title="LSTAT DISTRIBUTION", color=bar.index)

In [119]:
px.box(train['lstat'], title = "BOXPLOT DIAGRAM OF LSTAT VARIABLE WITH OUTLIERS")

Se debe generar un dataframe auxiliar para así eliminar temporalmente los valores NaN y encontrar la moda que en este caso nos servirá para imputar los datos faltantes más adelante.

In [120]:
dataframeLstat = generateDataframe(train.copy(), 'lstat')
modeLstat = dataframeLstat['lstat'].mode()[0]
print('Mode found:' , modeLstat)

Q1 6.58
Q3 14.27
IQR 7.6899999999999995
Upper bound: (array([  22,   31,   52,  148,  231,  265,  279,  386,  439,  575,  596,
        827,  842,  873,  874,  951,  989, 1106, 1183]),)
Lower bound: (array([], dtype=int64),)
Mode found: 1.73


Columna LSTAT en el dataset de entrenamiento con valores NaN.

In [121]:
train['lstat'].describe()

count     1221
unique     166
top       1.73
freq        21
Name: lstat, dtype: object

Columna LSTAT en el dataset de entrenamiento con valores imputados dada la moda.

In [122]:
train['lstat'].fillna(modeLstat, inplace = True)
train['lstat'] = train['lstat'].astype('float64')
train['lstat'].describe()

count    1296.000000
mean       10.343326
std         5.778513
min         1.730000
25%         5.890000
50%         9.590000
75%        14.100000
max        34.410000
Name: lstat, dtype: float64

Se borran los outliers en el dataset de entrenamiento.

In [123]:
train.reset_index(inplace = True, drop = True)
train = deleteOutliersSetTrain(train.copy(), 'lstat')

Q1 5.89
Q3 14.1
IQR 8.21
Upper bound: (array([  27,   38,   59,  161,  253,  289,  304,  418,  475,  621,  642,
        882,  898,  929,  930, 1010, 1054, 1176, 1256]),)
Lower bound: (array([], dtype=int64),)


In [124]:
px.box(train['lstat'], title = "BOXPLOT DIAGRAM OF LSTAT")

Se verifica la cantidad de valores faltantes para la variable LSTAT.

In [125]:
train['lstat'].isna().sum()

0

In [126]:
train.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1277 entries, 0 to 1276
Data columns (total 14 columns):
 #   Column   Non-Null Count  Dtype  
---  ------   --------------  -----  
 0   zn       1277 non-null   float64
 1   black    1277 non-null   float64
 2   rad      1277 non-null   object 
 3   indus    1277 non-null   float64
 4   ptratio  1277 non-null   object 
 5   lstat    1277 non-null   float64
 6   chas     1277 non-null   float64
 7   crim     1277 non-null   float64
 8   dis      1277 non-null   float64
 9   tax      1277 non-null   object 
 10  rm       1277 non-null   float64
 11  age      1277 non-null   object 
 12  nox      1277 non-null   float64
 13  medv     1160 non-null   object 
dtypes: float64(9), object(5)
memory usage: 139.8+ KB


## Remover Datos Duplicados Exactos

In [127]:
# X_train.duplicated().sum()

In [128]:
# X_train.drop_duplicates(keep = False, inplace = True, ignore_index = True)

## Perfilamiento del set de entrenamiento - reporte 2

In [129]:
# profile = ProfileReport(X_train, title="Boston House Prices Report 2",
#                         explorative=True, samples=None, correlations=None, 
#                         missing_diagrams=None, duplicates=None, interactions=None)

In [130]:
# profile.to_notebook_iframe()   

In [131]:
# profile.to_file("../reports/Boston_house_prices_report_2.html")

## Identificacion de Variables
- Variables de entrada y de salida
- Tipo de Variables (categoricas o Numericas)
- Tipo de datos (int, float, string, factor, boolean, ...)

## Procesamiento de Datos Faltantes
Las opciones que se pueden usar dependiendo del analisis de los datos son:
### Borrar Filas
- Borrar las filas que les falten todos los datos
- Borrar Solo las filas que les falta las variables mas importantes o la salida

### Reemplazar datos faltantes con la Media/ Moda/ Mediana (Mean/ Mode/ Median Imputation)
De necesitarse hacer imputacion se debe verificar los outliers para no cometer errores en la imputacion

## Remover Datos Duplicados Exactos

Hacer al principio luego de cargar los datos y cada vez luego de Hacer imputaciones, eliminacion de columnas o registros.

## Analisis Univariable

Estadistico Descriptico y Analisis

### Variables Numericas

Tendencia Central  | Medida de Dispersión  |  Visualizacion
 :---: | :---: | :---:
Media   | Rango      |  Histogramas
Mediana | Cuartiles  |  Boxplots
Moda    | Rango inter cuartil (IQR)|  
Minimo  | Varianza   |  
Maximo  | Desviacion Estandard | 
   .    | Skewness   |  
 .      | Kurtosis   |  

### Variables Categoricas
- Numero de elementos por categoria
- Porcentaje de elementos por categoria
- Graficos de barras

## Analisis Bivariable
Estadistico Descriptico y Analisis

### Numericas vs Numericas
- Scatter Plot
- Heatmap
- Correlacion

### Categoricas vs Categoricas

- Two-way table
- Graficas de Barras apiladas
- Chi-Square Test

### Categoricas vs Numericas
- Z-Test/ T-Test
- ANOVA

## Procesamiento de Outliers
### Deteccion de Outliers (Univariables y Bi variables)
- Boxplots
- Scatter Plots
- Histogramas

### Remover Outliers
las opciones son:
- Borrar los datos atipicos (outliers) 
- Transformar la variable (Ej: escalar, cambiar a esccala Log o a Escala lineal)
- Reemplazar los outliers conla Media/ Moda/ Mediana (Mean/ Mode/ Median Imputation)
- Reemplazar valores con modelo predictivo
- Separarlos y analizarlos aparte (si son muchos outliers)


## Procesamiento de Outliers
### Deteccion de Outliers (Univariables y Bi variables)
- Boxplots
- Scatter Plots
- Histogramas

### Remover Outliers
las opciones son:
- Borrar los datos atipicos (outliers) 
- Transformar la variable (Ej: escalar, cambiar a esccala Log o a Escala lineal)
- Reemplazar los outliers conla Media/ Moda/ Mediana (Mean/ Mode/ Median Imputation)
- Reemplazar valores con modelo predictivo
- Separarlos y analizarlos aparte (si son muchos outliers)


## Feature Engineering 
### Transformacion de Variables
**Se usa cuando es necesario:**
- Cambiar la escala las variables (Normalizar, escalamiento, etc), esto no cambia la forma de la distribucion de los datos
- Cambiar relaciones No ineales en Lineales (Ej: cambiar la escala Logaritmica a lineal)
- Usar una la distribución simétrica en vez de una distribución sesgada o asimetrica, Para una la distribución sesgada a la derecha, tomamos la raíz cuadrada / cúbica o el logaritmo de la variable, y para la desviación a la izquierda, tomamos el cuadrado / cubo o exponencial de las variables.
- Cambair variables continuas a categoricas

Algunas transformaciones son:

#### Normalizacion

#### Escalamiento

#### Logaritmica

#### raíz cuadrada / cúbica

#### Binning , Cambios de Numericas a Categoricas

## Analisis Univariable y Bivariable Final
Luego de Realizar:
- Tratamiento de datos Nulos
- Tratamiento de outliers
- Transformacion de Variables

Es necesario realizar nuevamente analisis univarible y bivariable para identificar como cambiaron nuestros datos y para verificar si estan listos para usarse para crear un modelo.

### Creacion de Variables
pueden ser:
#### Crear Variables derivadas de Otras

#### Crear Variables de Categorico a Numerico
Puede ser
- Definir un numero a cada categoria
- One hot encoder (dummy columns)

## Reduccion de Dimensionalidad y Seleccion de Variables (PCA)

De ser necesario si tiene muchas variables o sospecha que aun tiene variables que no aportan mucho

# MODELAMIENTO DE LOS DATOS (MACHINE LEARNING)


# Validacion y Evaluacion Cruzada (k-fold Cross Validation)

Se hace seleccion de los mejores modelos usando el Training Set y k-fold Cross Validation

# Optimizacion de Hiper parametros (Hyper Parameter optimization)

Se seleccionan solo los mejores modelos para realizar el ajuste de hiperparametros, ya que tiene una carga computacional alta.

Al final se obtienen los parametros del mejor modelo

# Evaluacion final del modelo con el Test set

Tomar los parametros obtenidos en el paso anterior, se crea el modelo con esos pararmetros y se entrena el modelo con todos los datos del **Train -set**

Finalmente se realiza la evaluacion (segun su problema si es de regresion o de clasificacion) usando el **Test - set** para definir si el modelo obtenido esta bien. Compare los resultados con el **Train -set** vs los resultados con el **Test - set**

# Implementacion del Modelo (Deploying)
Con el análisis básico y el ajuste hecho, comienza el trabajo real (ingeniería).

El último paso para poner en produccion el modelo de prediccion sera:
1. Entrenarlo en todo el conjunto de datos nuevamente, para hacer un uso completo de todos los datos disponibles. 
2. Usar los mejores parámetros encontrados mediante la validación cruzada, por supuesto. Esto es muy similar a lo que hicimos al principio, pero esta vez teniendo una idea de su comportamiento y estabilidad. La evaluación se realizó con honestidad, en divisiones distintas de entrenamiento / prueba.

El predictor final se puede serializar y grabar en el disco, de modo que la próxima vez que lo usemos, podemos omitir todo el entrenamiento y usar el modelo capacitado directamente:

In [132]:
#import pickle # Esta es una libreria de serializacion nativa de python, puede tener problemas de seguridad
from joblib import dump # libreria de serializacion

# garbar el modelo en un archivo
#dump(Modelo_final, 'Nombre_Archivo_Modelo.joblib')

# Comunicacion de Resultados (Data Story Telling)

# Conclusiones

# Ayudas Y Referencias

- https://medium.com/@joserzapata/paso-a-paso-en-un-proyecto-machine-learning-bcdd0939d387
- [Proyecto de Principio a Final sobre readmision de pacientes con Diabetes](https://github.com/JoseRZapata/Readmission-ML-Project)

- [a-complete-machine-learning-walk-through-in-python-part-one](https://towardsdatascience.com/a-complete-machine-learning-walk-through-in-python-part-one-c62152f39420)


- [a-starter-pack-to-exploratory-data-analysis-with-python-pandas-seaborn-and-scikit-learn](https://towardsdatascience.com/a-starter-pack-to-exploratory-data-analysis-with-python-pandas-seaborn-and-scikit-learn-a77889485baf#249d)

- [a-data-science-for-good-machine-learning-project-walk-through-in-python-part-one](https://towardsdatascience.com/a-data-science-for-good-machine-learning-project-walk-through-in-python-part-one-1977dd701dbc)

- [Ejemplos de Kaggle](https://www.kaggle.com/kernels?sortBy=hotness&group=everyone&pageSize=20&language=Python&kernelType=Notebook)

- [END to END ML from data colletion to deployment](https://medium.com/datadriveninvestor/end-to-end-machine-learning-from-data-collection-to-deployment-ce74f51ca203)

Docente: [Jose R. Zapata](https://joserzapata.github.io)
- https://joserzapata.github.io
- https://twitter.com/joserzapata
- https://www.linkedin.com/in/jose-ricardo-zapata-gonzalez/   